File: frontend/lib/types/reservation.ts
export type ReservationType = 'hold' | 'trade_in' | 'offer_accepted';

export type ReservationStatus =
  | 'pending'
  | 'deposit_paid'
  | 'completed'
  | 'cancelled'
  | 'expired';

export type CancellationReason =
  | 'customer_backed_out'
  | 'trade_fell_through'
  | 'expired'
  | 'sold_elsewhere'
  | 'other';

/** Admin-selectable types. `offer_accepted` is created by the offer flow, not by hand. */
export const RESERVATION_TYPE_OPTIONS: { value: ReservationType; label: string }[] = [
  { value: 'hold', label: 'Hold' },
  { value: 'trade_in', label: 'Trade-In' },
];

export const RESERVATION_STATUS_OPTIONS: { value: ReservationStatus; label: string }[] = [
  { value: 'pending', label: 'Pending' },
  { value: 'deposit_paid', label: 'Deposit Paid' },
  { value: 'completed', label: 'Completed' },
  { value: 'cancelled', label: 'Cancelled' },
  { value: 'expired', label: 'Expired' },
];

export const CANCELLATION_REASON_OPTIONS: { value: CancellationReason; label: string }[] = [
  { value: 'customer_backed_out', label: 'Customer backed out' },
  { value: 'trade_fell_through', label: 'Trade fell through' },
  { value: 'expired', label: 'Expired' },
  { value: 'sold_elsewhere', label: 'Sold elsewhere' },
  { value: 'other', label: 'Other' },
];

/** Default hold length per type, mirroring the server. */
export function defaultExpiryDays(type: ReservationType): number {
  if (type === 'trade_in') return 30;
  if (type === 'offer_accepted') return 3;
  return 7;
}

/** Admin view. Includes the holder's identity — never render this on a public page. */
export interface AdminReservation {
  id: string;
  listing_id: string;
  listing_title: string;
  listing_image: string | null;
  listing_price: number | null;
  listing_sold: boolean;

  type: ReservationType;
  type_label: string;
  status: ReservationStatus;
  status_label: string;

  user_id: string | null;
  user_name: string | null;
  user_email: string | null;
  is_unassigned: boolean;
  user_missing: boolean;

  agreed_price: number;
  trade_in_credit: number;
  deposit_required: boolean;
  deposit_amount: number;
  deposit_refundable: boolean;
  deposit_paid_amount: number;
  deposit_paid_at: string | null;
  deposit_payment_method: string | null;
  deposit_order_id: string | null;
  final_order_id: string | null;
  balance_due: number;
  is_over_credited: boolean;

  expires_at: string | null;
  is_expired: boolean;
  internal_note: string | null;
  cancellation_reason: string | null;
  needs_review: boolean;
  needs_review_reason: string | null;
  source_conversation_id: string | null;

  created_at: string;
  updated_at: string;
  completed_at: string | null;
}

export interface ReservationSummary {
  active_holds: number;
  deposits_held: number;
  expiring_48h: number;
  needs_review: number;
  unassigned: number;
}

/** Customer-facing view. Contains no identity information and no admin note. */
export interface MyReservation {
  id: string;
  type: ReservationType;
  type_label: string;
  status: ReservationStatus;
  status_label: string;
  agreed_price: number;
  trade_in_credit: number;
  deposit_required: boolean;
  deposit_amount: number;
  deposit_refundable: boolean;
  deposit_paid_amount: number;
  deposit_paid_at: string | null;
  balance_due: number;
  expires_at: string | null;
  currency: string;
}

/**
 * What /api/reservations/listing/{id} returns. Non-holders get only `is_reserved`,
 * `badge` and `message` — never who the holder is.
 */
export interface ListingReservationState {
  is_reserved: boolean;
  is_mine?: boolean;
  type?: ReservationType;
  badge?: string;
  message?: string;
  reservation?: MyReservation;
}

export interface DepositDetails {
  reservation_id: string;
  listing_id: string;
  listing_title: string;
  listing_image: string | null;
  currency: string;
  line_item_label: string;
  agreed_price: number;
  trade_in_credit: number;
  deposit_amount: number;
  deposit_refundable: boolean;
  balance_after_deposit: number;
  expires_at: string | null;
  shipping_charged: boolean;
  tax_charged: boolean;
}

export interface CreateReservationPayload {
  listingId: string;
  userId: string;
  type: ReservationType;
  agreedPrice: number;
  tradeInCredit: number;
  depositRequired: boolean;
  depositAmount: number;
  depositRefundable: boolean;
  expiresAt: string | null;
  noExpiration: boolean;
  internalNote: string | null;
}

export interface UpdateReservationPayload {
  userId?: string;
  agreedPrice?: number;
  tradeInCredit?: number;
  depositRequired?: boolean;
  depositAmount?: number;
  depositRefundable?: boolean;
  expiresAt?: string | null;
  noExpiration?: boolean;
  internalNote?: string | null;
}

// ---------- display helpers ----------

export function statusBadgeClasses(status: ReservationStatus): string {
  switch (status) {
    case 'pending':
      return 'bg-amber-100 text-amber-800';
    case 'deposit_paid':
      return 'bg-green-100 text-green-800';
    case 'expired':
      return 'bg-gray-200 text-gray-700';
    case 'cancelled':
      return 'bg-red-100 text-red-700';
    case 'completed':
      return 'bg-blue-100 text-blue-800';
    default:
      return 'bg-gray-100 text-gray-700';
  }
}

export function typeBadgeClasses(type: ReservationType): string {
  switch (type) {
    case 'trade_in':
      return 'bg-purple-100 text-purple-800';
    case 'offer_accepted':
      return 'bg-sky-100 text-sky-800';
    default:
      return 'bg-slate-100 text-slate-700';
  }
}

/** Hours until expiry; null when there is no expiration. */
export function hoursUntil(expiresAt: string | null): number | null {
  if (!expiresAt) return null;
  return (new Date(expiresAt).getTime() - Date.now()) / 36e5;
}

/** True when a hold lapses in under 48 hours — drives the red highlight. */
export function isExpiringSoon(expiresAt: string | null): boolean {
  const hours = hoursUntil(expiresAt);
  return hours !== null && hours > 0 && hours < 48;
}

/** "in 6 days" / "in 5 hours" / "expired". */
export function relativeExpiry(expiresAt: string | null): string {
  if (!expiresAt) return 'No expiration';

  const hours = hoursUntil(expiresAt)!;
  if (hours <= 0) return 'Expired';
  if (hours < 1) return `in ${Math.max(1, Math.round(hours * 60))} min`;
  if (hours < 48) return `in ${Math.round(hours)} hour${Math.round(hours) === 1 ? '' : 's'}`;

  const days = Math.round(hours / 24);
  return `in ${days} day${days === 1 ? '' : 's'}`;
}

export function formatCurrency(amount: number, currency = 'USD'): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
    minimumFractionDigits: 2,
  }).format(amount);
}

File: backend/GuitarDb.API/Models/Reservation.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace GuitarDb.API.Models;

/// <summary>
/// A reservation is the single source of truth for "who is this guitar promised to".
/// One active reservation per listing, max. Replaces the old MyListing.Pending flag
/// and backs the accepted-offer cart lock.
/// </summary>
public class Reservation
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    [BsonIgnoreIfNull]
    public string? Id { get; set; }

    [BsonElement("listing_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string ListingId { get; set; } = string.Empty;

    /// <summary>
    /// The single customer allowed to buy this guitar while the reservation is active.
    /// Null means "migrated but unassigned" — blocks checkout for everyone (safe default).
    /// </summary>
    [BsonElement("user_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    [BsonIgnoreIfNull]
    public string? UserId { get; set; }

    [BsonElement("type")]
    public string Type { get; set; } = ReservationType.Hold;

    [BsonElement("status")]
    public string Status { get; set; } = ReservationStatus.Pending;

    /// <summary>
    /// Terms are locked at creation. Balance is always computed from this, never from
    /// the live listing price, so editing the listing later does not move the goalposts.
    /// </summary>
    [BsonElement("agreed_price")]
    public decimal AgreedPrice { get; set; }

    [BsonElement("trade_in_credit")]
    public decimal TradeInCredit { get; set; } = 0m;

    [BsonElement("deposit_required")]
    public bool DepositRequired { get; set; } = false;

    [BsonElement("deposit_amount")]
    public decimal DepositAmount { get; set; } = 0m;

    [BsonElement("deposit_refundable")]
    public bool DepositRefundable { get; set; } = false;

    [BsonElement("deposit_paid_amount")]
    public decimal DepositPaidAmount { get; set; } = 0m;

    [BsonElement("deposit_paid_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    [BsonIgnoreIfNull]
    public DateTime? DepositPaidAt { get; set; }

    /// <summary>"card", "paypal", or free text for manual (cash / Venmo / Zelle).</summary>
    [BsonElement("deposit_payment_method")]
    [BsonIgnoreIfNull]
    public string? DepositPaymentMethod { get; set; }

    /// <summary>Order id of the deposit order, so the paper trail is complete.</summary>
    [BsonElement("deposit_order_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    [BsonIgnoreIfNull]
    public string? DepositOrderId { get; set; }

    /// <summary>Order id of the final balance order.</summary>
    [BsonElement("final_order_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    [BsonIgnoreIfNull]
    public string? FinalOrderId { get; set; }

    /// <summary>Null = no expiration.</summary>
    [BsonElement("expires_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    [BsonIgnoreIfNull]
    public DateTime? ExpiresAt { get; set; }

    /// <summary>Admin-only. Never returned on any public or customer-facing endpoint.</summary>
    [BsonElement("internal_note")]
    [BsonIgnoreIfNull]
    public string? InternalNote { get; set; }

    /// <summary>Set when cancelled. Dropdown reason + optional free text.</summary>
    [BsonElement("cancellation_reason")]
    [BsonIgnoreIfNull]
    public string? CancellationReason { get; set; }

    /// <summary>
    /// Set when a DepositPaid reservation runs past expiry. Money changed hands, so we
    /// never auto-release — we flag it and let the admin decide.
    /// </summary>
    [BsonElement("needs_review")]
    public bool NeedsReview { get; set; } = false;

    [BsonElement("needs_review_reason")]
    [BsonIgnoreIfNull]
    public string? NeedsReviewReason { get; set; }

    /// <summary>Set once the 48h "expiring soon" warning has gone out, so it only fires once.</summary>
    [BsonElement("expiring_soon_notified_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    [BsonIgnoreIfNull]
    public DateTime? ExpiringSoonNotifiedAt { get; set; }

    /// <summary>Conversation id when this reservation came from an accepted offer.</summary>
    [BsonElement("source_conversation_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    [BsonIgnoreIfNull]
    public string? SourceConversationId { get; set; }

    [BsonElement("created_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("updated_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("completed_at")]
    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    [BsonIgnoreIfNull]
    public DateTime? CompletedAt { get; set; }

    // ---- Derived values. Never persisted; always recomputed server-side. ----

    /// <summary>
    /// What the reserved user still owes: agreed price - deposit paid - trade-in credit.
    /// Never negative. Recomputed on every read so a stale document can't leak a wrong price.
    /// </summary>
    [BsonIgnore]
    public decimal BalanceDue =>
        Math.Max(0m, AgreedPrice - DepositPaidAmount - TradeInCredit);

    /// <summary>True when credits exceed the price — flagged for admin review, never auto-refunded.</summary>
    [BsonIgnore]
    public bool IsOverCredited =>
        (DepositPaidAmount + TradeInCredit) > AgreedPrice;

    /// <summary>Active = still blocking the listing for everyone but the reserved user.</summary>
    [BsonIgnore]
    public bool IsActive =>
        Status == ReservationStatus.Pending || Status == ReservationStatus.DepositPaid;

    [BsonIgnore]
    public bool IsExpired =>
        ExpiresAt.HasValue && ExpiresAt.Value <= DateTime.UtcNow;

    /// <summary>
    /// Unassigned migrated reservations block checkout for everyone until the admin
    /// assigns the right customer.
    /// </summary>
    [BsonIgnore]
    public bool IsUnassigned => string.IsNullOrEmpty(UserId);
}

public static class ReservationType
{
    public const string Hold = "hold";
    public const string TradeIn = "trade_in";
    public const string OfferAccepted = "offer_accepted";

    public static readonly string[] All = { Hold, TradeIn, OfferAccepted };

    /// <summary>Admin-selectable types. OfferAccepted is created by the offer flow, not by hand.</summary>
    public static readonly string[] AdminSelectable = { Hold, TradeIn };

    public static bool IsValid(string? type) =>
        !string.IsNullOrEmpty(type) && All.Contains(type);

    public static string Label(string type) => type switch
    {
        Hold => "Hold",
        TradeIn => "Trade-In",
        OfferAccepted => "Accepted Offer",
        _ => type
    };

    /// <summary>Public-facing badge. Must never hint at who the holder is.</summary>
    public static string PublicBadge(string type) => type switch
    {
        TradeIn => "Pending Trade-In",
        _ => "On Hold"
    };

    /// <summary>Default hold length per type, per spec §5.</summary>
    public static int DefaultExpiryDays(string type) => type switch
    {
        TradeIn => 30,
        OfferAccepted => 3,
        _ => 7
    };
}

public static class ReservationStatus
{
    public const string Pending = "pending";
    public const string DepositPaid = "deposit_paid";
    public const string Completed = "completed";
    public const string Cancelled = "cancelled";
    public const string Expired = "expired";

    public static readonly string[] All = { Pending, DepositPaid, Completed, Cancelled, Expired };

    /// <summary>Statuses that still reserve the listing.</summary>
    public static readonly string[] Active = { Pending, DepositPaid };

    public static bool IsValid(string? status) =>
        !string.IsNullOrEmpty(status) && All.Contains(status);

    public static string Label(string status) => status switch
    {
        Pending => "Pending",
        DepositPaid => "Deposit Paid",
        Completed => "Completed",
        Cancelled => "Cancelled",
        Expired => "Expired",
        _ => status
    };
}

public static class ReservationCancellationReason
{
    public const string CustomerBackedOut = "customer_backed_out";
    public const string TradeFellThrough = "trade_fell_through";
    public const string Expired = "expired";
    public const string SoldElsewhere = "sold_elsewhere";
    public const string Other = "other";

    public static readonly string[] All =
        { CustomerBackedOut, TradeFellThrough, Expired, SoldElsewhere, Other };

    public static bool IsValid(string? reason) =>
        !string.IsNullOrEmpty(reason) && All.Contains(reason);
}
